### DISCUSSION TITLE
split_map versions

### DISCUSSION THREAD

**Discussion by lopezvoliver** [opengeos/leafmap#680] [Ideas]
There are three versions of `split_map`: 

(1) as a method of `leafmap.Map` (ipyleaflet.Map)
(2) as function `leafmap.split_map`
(3) as a method of `leafmap.foliumap.Map` (folium.Map)

> The `leafmap.foliumap.split_map` is not yet implemented (raises NotImplementedError)

The notebook `49_split_control.ipynb` shows how to use (3) because ipyleaflet SplitControl does not support streamlit, and it shows how to do a splitmap using:

- basemaps
- Cloud-optimized geotiff (COG) urls
- folium.WmsTileLayer 

The first two (basemap and COG urls) are easy to use, since you can pretty much just do `m.split_map(left_layer, right_layer)`, while for the third example you need to first setup each layer by calling `folium.WmsTileLayer`. This is because all versions of `split_map` first check if they are strings (a basemap, a url, etc) or an ipyleaflet.TileLayer. 

I am interested in making a `split_map` to compare between two local rasters, but I would like to know if there's room for improvement for streamlining the 3 versions of `split_map` and if there is already an effort to do so. I made some modifications to (1) that borrow code from `add_raster` (a call to `array_to_image` and `get_local_tile_layer`), and requires passing some keyword arguments to at least `get_local_tile_layer`, e.g., looks like this:

```python
import leafmap # ipyleaflet
m = leafmap.Map(center=[37.6,-119], zoom=9)
m.split_map(dem_ds, dem_class,
            left_args={"layer_name":"DEM", "colormap": "terrain", 
                       "visible":True, 
                       "port":54605,
            },
            right_args={"layer_name":"Classified DEM","colormap": "coolwarm", 
                        "visible":True,
            },
           # left_array_args ={}   # not needed in this example but could be passed to array_to_image
           # right_array_args ={}   # not needed in this example but could be passed to array_to_image
            )
m
```

![demo_split_map](https://github.com/opengeos/leafmap/assets/14804652/eb13046d-e5dc-4a80-b053-d15cdf721250)

Similarly, I think the `folium.WmsTileLayer` example could be simplified by using `left_args` and `right_args`. 

However, these changes would have to be repeated for all versions of `split_map`. Should we add these incrementally (e.g. I could make a PR for the changes to (1))?

**Comment by lopezvoliver** [comment]
I also wanted to check the reason for disabling dragging in (1) and (2)? could we add an option for this instead?

**Comment by giswqs** [comment]
These are great improvements. Please feel free to submit a PR. 

Dragging is temporarily disabled due to the ipylealfet bug https://github.com/jupyter-widgets/ipyleaflet/issues/1115. I don't think it has been fixed. You can give it a try by setting `dragging` to `True`

**Reply by lopezvoliver** [reply]
![demo_split_map_drag](https://github.com/opengeos/leafmap/assets/14804652/06d99342-c442-4c78-a22a-c882840e2280)


I didn't experience the bug, but since it's easy to set `dragging=True` after split_map, I didn't modify it in `split_map` for now. 

Edit: I replaced the animation to the correct one I wanted to show.

**Reply by giswqs** [reply]
dragging is for dragging the map, not the slider. Try setting `dragging=True` and you may experience the bug.

**Reply by lopezvoliver** [reply]
I originally attached the wrong animation. I edited my comment above with the correct one. 

However, for folium maps, I'm not sure what the expected behavior is? See below: the split slider stays in the center, and the whole map drags around it

![split_map_folium_no_drag](https://github.com/opengeos/leafmap/assets/14804652/26244081-5762-43ca-99d2-ec1c98b28b4c)

Setting `m.dragging=True` makes no difference in this case.

**Reply by giswqs** [reply]
It seems the dragging issue has been resloved upstream. We can restore the dragging option now.

**Reply by giswqs** [reply]
I thought folium has fixed the issue, but it is happening again. 
https://github.com/python-visualization/folium/issues/1730

**Comment by lopezvoliver** [comment]
#681 Added support for numpy.ndarray and xarray.DataArray in split_map versions (1) and (3) mentioned above.
